### DISCUSSION TITLE
Mermaid UX Can Be Poor

### DISCUSSION THREAD

**Discussion by fuzyll** [outline/outline#5834] [General] [answered]
Tried to sit down and use the Mermaid integration today on our self-hosted instance (v0.71.0). Experience was less-than-stellar. I have absolutely no idea what, of the things below, is actionable, which is why I've made this a discussion. I think ultimately a lot of these are Mermaid's issues, but it might be possible to make some changes to mitigate them in Outline.

All I wanted to do today was create a timeline. Unfortunately, the version of Mermaid that Outline uses (v9.3.0) is either too old to have it or doesn't have experimental features turned on. (The [Mermaid docs](https://mermaid.js.org/syntax/timeline.html), which don't appear to have an option to view older docs, say the *syntax* is stable, but *icons* aren't, and doesn't say when support was added at all. So, I have no idea what the actual issue is. I'm just making an educated guess.)

Not a problem - I can still do what I want with a Gantt chart.

First issue: When you make a syntax error, the page gets completely messed up because Mermaid helpfully renders this huge thing on the right (which is how I know what version is running):

<img width="706" alt="image" src="https://github.com/outline/outline/assets/607452/494d3b71-aa85-4b62-95ed-efd7111c3c06">

Not a *big* deal, just very annoying. I also ran into what I would consider is a display issue with milestones, but that's absolutely Mermaid's problem (and still does, I would argue, the wrong thing on v10.4.1).

As I added things, I needed to change a date from 2023-mm-yy to 2024-mm-yy. So, I deleted the '3'. And then the tab crashed, because the Mermaid preview decided it would instantly try and render a Gantt chart from the year 202 to the year 2023. Opening this tab up again just caused the same behavior.

Since I couldn't find a way to prevent Mermaid from running, I considered the entire document a loss. I downloaded it as markdown, deleted it, copy/pasted the other content onto a new page and copy/pasted the Mermaid diagram (with the date fixed) onto a different page where I could edit it and limit the blast radius. Later, I accidentally added an extra '4' to a date (20244) and I was back to the same behavior.

I'm writing this after having given up, deleted my new test document, and embedded a Google Docs spreadsheet instead so I can move on with my life. It looks horrible and is far harder to modify, but at least the information is presented *somewhat* visually now.

I've confirmed that [Mermaid Live](https://mermaid.live/) does the same thing wrt the smaller year (202) without crashing, but just doesn't display anything for the larger year (20244). Here's a cleaned up version of the chart that will trigger the issue:

```mermaidjs
gantt
  title Milestone Test
  section Milestones
    Milestone 1 : milestone, m1, 2023-09-18, 7d
    Milestone 2 : milestone, m2, after m1, 28d
    Milestone 3 : milestone, m3, after m2, 7d
  section Tasks
    Task 1 : a1, 2023-09-18, 7d
    Task 2 : a2, after a1, 28d
    Task 3 : after a2, 7d
```

Just change `m1`'s year to see the issue.

**Comment by tommoor** [comment, marked as answer]
Mermaid is somewhat stuck on `9.3.0` because versions above that ballooned in size, upgrading to 10 literally makes the _entire_ Outline codebase 30% larger which is just unacceptable.

I'm trying to get to the bottom of why you're seeing that error screen with the bomb, it shouldn't show up and is clearly messing with the page layout too…

**Reply by tommoor** [reply]
Okay, the Syntax Error appearing at the side of the document was fixed in 091ef340f.

I think the speed of rendering with dates far apart should be brought up on the parent repo and fixed there, as you mention it's upstream and there isn't a lot we can do.

**Reply by fuzyll** [reply]
Thank you! And I agree, I'll go file an issue over there.

EDIT: Found an existing issue. And, ironically, you already commented on it! It's almost a year old at this point, though...

**Comment by routerino** [comment]
> Mermaid is somewhat stuck on `9.3.0` because versions above that ballooned in size, upgrading to 10 literally makes the _entire_ Outline codebase 30% larger which is just unacceptable.
> 
> I'm trying to get to the bottom of why you're seeing that error screen with the bomb, it shouldn't show up and is clearly messing with the page layout too…

If you are interested in dropping the mermaid support (especially given it can break articles irrecoverably if updated wrong, this has happened to me too), Diagrams.net has full mermaid support. Wouldn't be 100% as nice (for mermaid), but if you were to mature the diagrams.net integration (as per https://github.com/outline/outline/issues/2831) you could ditch the mermaid integration entirely and just make it available via diagrams.net.

**Comment by antran22** [comment]
@fuzyll I have opened two PRs for those issues. #5852 deal with the rendering flow. #5853 updating Mermaid to the newest version.

Right now the new update to Mermaid 10.4.0 is being discussed because of the issues with regard to bundle size. However, the fix at #5852 can be included first, improving the experience with Mermaid in Outline.

I have built a patched Docker image that I'm deploying for my team now with those two PRs. If you feel adventurous you can try it on your own instance. [Docker Hub Link](https://hub.docker.com/layers/antran2202/outline/0.71.0-patch3/images/sha256-f339692c07b5d7f47771ed6d63f82e855b9d2b9c2e92d98819e90b8eabf7ced0?context=explore)

**Comment by zenoxygenai** [comment]
Any chance mermaid will be updated?

**Reply by tommoor** [reply]
It has been, we're using the latest version now
